File: DataStructureAndAlgorithm/Graph/BellmanFord/C++/BellmanFord.cpp
#include "BellmanFord.h"
#include <iostream>

void bellmanFord(int startNode, int numVertices, const std::vector<Edge>& edges) {
    std::vector<int> dist(numVertices, std::numeric_limits<int>::max());
    dist[startNode] = 0;

    // Relax all edges |V| - 1 times.
    for (int i = 0; i < numVertices - 1; ++i) {
        for (const Edge& edge : edges) {
            if (dist[edge.src] != std::numeric_limits<int>::max() &&
                dist[edge.src] + edge.weight < dist[edge.dest]) {
                dist[edge.dest] = dist[edge.src] + edge.weight;
            }
        }
    }

    // Check for negative-weight cycles.
    for (const Edge& edge : edges) {
        if (dist[edge.src] != std::numeric_limits<int>::max() &&
            dist[edge.src] + edge.weight < dist[edge.dest]) {
            std::cout << "Graph contains negative weight cycle" << std::endl;
            return;
        }
    }

    std::cout << "Vertex   Distance from Source" << std::endl;
    for (int i = 0; i < numVertices; ++i) {
        std::cout << i << "\t\t" << dist[i] << std::endl;
    }
}

File: DataStructureAndAlgorithm/Graph/BellmanFord/Python/bellman_ford.py
def bellman_ford(graph, start_node):
    # graph is a dictionary where keys are nodes and values are lists of (neighbor, weight) tuples
    # Example: graph = {0: [(1, 4), (2, 2)], 1: [(3, 2)], 2: [(1, 1), (3, 5)], 3: []}

    distances = {node: float('inf') for node in graph}
    distances[start_node] = 0

    # Relax edges |V| - 1 times
    for _ in range(len(graph) - 1):
        for node in graph:
            for neighbor, weight in graph[node]:
                if distances[node] != float('inf') and distances[node] + weight < distances[neighbor]:
                    distances[neighbor] = distances[node] + weight

    # Check for negative cycles
    for node in graph:
        for neighbor, weight in graph[node]:
            if distances[node] != float('inf') and distances[node] + weight < distances[neighbor]:
                print("Graph contains negative weight cycle")
                return None

    return distances
